### PR TITLE
require `_EXECUTION_NAME` in payload

### PR DESCRIPTION
**Pull request by natebrennand**
If `sfncli` encounters a workflow payload that lacks `_EXECUTION_NAME` *and* uses the `workdirectory` flag, the environment variables are not provided.

This PR changes behavior to:
1. require that `_EXECUTION_NAME` is provided in the payload
2. always provide the full set of OS environment variables + `_EXECUTION_NAME` for the executing process.

----------
----------
----------

## Prior PR text

If `sfncli` encounters a workflow payload that lacks `_EXECUTION_NAME`,
the environment variables are not provided.

When not hitting the `if executionName != nil` codepath, `t.execCmd.Env`
would default to nil and then execute the process with either:

1. no environment variables
2. only the `WORK_DIR` env var

------

I think this should maybe be paired with a change to `workflow-manager`
to re-insert `_EXECUTION_NAME` to payloads whenever it is dropped by a
worker. Currently this is done by `sfncli`, but it relies on the value
being present in the input.

### CODE REVIEW

**Comment by nathanleiby** [comment]
cc @mohit @rgarcia in case you have opinions

**Comment by nathanleiby** [comment]
Do we think of execution name as required or optional? if required, we should probably fail here with a clear error message

**Comment by mohit** [comment]
defering to @rgarcia.

**Comment by rgarcia** [comment]
👍 this is a bug

> I think this should maybe be paired with a change to workflow-manager to re-insert _EXECUTION_NAME to payloads whenever it is dropped by a worker. Currently this is done by sfncli, but it relies on the value being present in the input.

FWIW I've pondered this and don't know of any way to accomplish this without some very heavy handed insertion of states between every user-provided state.

I do think we could monitor/alert on lack of _EXECUTION_NAME in a job's input by adding a check in workflow-manager's update loop. This would at least create a feedback loop to fix state machines that don't pass along _EXECUTION_NAME

**Comment by natebrennand** [comment]
More context, the `_EXECUTION_NAME` attribute is being dropped in the payload when passing through a lambda worker.

I now understand why @rgarcia chose to use `input json.RawMessage` as the parameter for the `mp-locker` lambda.

------

I can buy this being required, but if we do that then `workflow-manager` should definitely be changed to re-insert it.

**Comment by rgarcia** [comment]
actually... is it a bug? From https://godoc.org/os/exec

```
    // If Env is nil, the new process uses the current process's
    // environment.
```

This is what the code was relying on before

**Comment by rgarcia** [comment]
Also I'd consider _EXECUTION_NAME required, since we depend on it for important things like linking to logs for a workflow in hubble

**Comment by natebrennand** [comment]
I guess I didn't actually test what I thought was case `1.`, if there is no `WORK_DIR` needed then it will have everything. But if you _are_ using `WORK_DIR` then you *do* only get `WORK_DIR` and no other env vars.

**Comment by natebrennand** [comment]
I've now updated the PR to require that `_EXECUTION_NAME` is set.

File: cmd/sfncli/runner.go [hunk 1]
```diff
@@ -65,12 +67,12 @@ func (t *TaskRunner) Process(ctx context.Context, args []string, input string) e
 		return t.sendTaskFailure(TaskFailureTaskInputNotJSON{input: input})
 	}
 
-	// convention: if the input contains _EXECUTION_NAME, pass it to the environment of the command
-	var executionName *string
-	if e, ok := taskInput["_EXECUTION_NAME"].(string); ok {
-		executionName = &e
-		t.logger.AddContext("execution_name", *executionName)
+	// convention: pass _EXECUTION_NAME into the environment of the command
```
Comment (rgarcia) [inline cmd/sfncli/runner.go]: maybe add something about _EXECUTION_NAME being required in the task input

File: cmd/sfncli/runner.go [hunk 2]
```diff
@@ -65,12 +67,12 @@ func (t *TaskRunner) Process(ctx context.Context, args []string, input string) e
...
+	executionName, ok := taskInput["_EXECUTION_NAME"].(string)
+	if !ok {
+		return t.sendTaskFailure(TaskFailureTaskInputMissingExecutionName{input: input})
 	}
+	t.logger.AddContext("execution_name", executionName)
```
Comment (rgarcia) [inline cmd/sfncli/runner.go]: nit: `execution-name`

File: cmd/sfncli/runner.go (hunk 1 above)
Comment (natebrennand) [reply cmd/sfncli/runner.go]: called out in the README & code w/ cc475191afef07723d40bb56264811fa0cb8acee

File: cmd/sfncli/runner.go (hunk 2 above)
Comment (natebrennand) [reply cmd/sfncli/runner.go]: Leaving this as is because it was with an underscore before

**Pull request merged**